### ISSUE TITLE
Why not a real alternative to Node?

### ISSUE THREAD

**Issue by mvalente** [node-forward/discussions#9]
Node is great and V8 is great. 

But why not take Mozilla's Spider/Eon/Odin/Monkey and create a real alternative to Node? Namely something that would be async/callback Node-compatible but have the alternative of sync-only API to better welcome people that are coming from PHP/Python/etc ?

**Comment by zzmp** [comment]
There is a sync API for most things - it's just not recommended for performance reasons. Do you want something more? Or just a different base to compete with V8?

**Comment by Fishrock123** [comment]
> But why not take Mozilla's Spider/Eon/Odin/Monkey

V8 was the fastest at the time, and I believe is still faster. Also, re-writing a new node would be a _ton_ of manpower.

> have the alternative of sync-only API

Node has Sync apis for lots of things, but JavaScript is designed in a way that makes a fully "sync" program very sub-par. (Javascript is single-threaded, etc.)

Generators in es6 will help make APIs nicer to use in a "sync" way though.

**Edit: Re, below:**

> If I had my way I'd remove everything with Sync in its name from the core API.

100% Agree.

**Comment by rvagg** [comment]
http://r.va.gg/presentations/love-the-async/

Because async is _why_ Node is so good at what it does, make it sync and you lose most of what makes Node so compelling--other than the JavaScript.

If I had my way I'd remove everything with `Sync` in its name from the core API.

Regarding alternative JS engines, I would expect this to develop over time, V8 is actually a huge pain in the backside to track for Node-core because of their development style and having alternatives would be great. It really takes someone with time on their hands to come up with something, ideally something that could be swapped out directly for Node or a way to abstract the parts of Node-core that are directly tied to V8's API.

**Comment by yoshuawuyts** [comment]
> If I had my way I'd remove everything with Sync in its name from the core API.

@rvagg Just for argument's sake, what would the impact of this be for both user land and maintaining core? I find the thought to be quite interesting to say the least.

**Comment by Qard** [comment]
Well, the entire require system wouldn't work, since it uses
fs.readFileSync().

**Comment by rvagg** [comment]
@Qard obviously you don't break sync `require()`, no reason it needs to also require a public fs sync API. This is just hypothetical and philosophical anyway; the main point being that I'd rather we put on-boarding effort into helping think in asynchronous terms instead of trying to munge our platforms and languages so that everything is sequential. Same goes for the false-promise of generators, you're just attempting to avoid async when it'd probably be better if you just accept that the world is async, everything we interact with in our programs is async and forcing synchronous behaviours is always going to be suboptimal because it means we're abstracting ourselves away from what's really happening by yet another step just so we can think in classic programming terms.

**Comment by mvalente** [comment]
No, Node doesnt have sync alternatives for everything.

Yes, I think that a different codebase (SpiderMonkey) would be important.

Yes, V8 was fast at the time. And its fast now. But current SpiderMonkey's versions are pretty fast as well.

I dont agree that rewriting a "new" node would be a ton of manpower. At least not for sync API, since SpiderMonkey has FFI as part of the code and it would be pretty easy to make the FFI interface available to JS.

JS is not singlethreaded. The engine is or isnt. And singlethreadedness hasnt stopped Apache and other webservers from being alternatives.

I undestand the extremist position of "remove everything with Sync". Great. Do understand that there are thousands of programmers (PHP, Python, etc) that find it hard to move to JS serverside because of all the async stuff and callback hell.

I know that async is what makes node fast. I've been around long enough ( http://goo.gl/5UpGwO ). But I find that the async mindset is too extremist. Async is good. Only async is bad. At least in terms of wider adoption of serverside JS.

Node is async and that is great because its fast. Nothing stops Spidermonkey being modified to also be async. And have sync as an alternative.

**Comment by domenic** [comment]
Async doesn't mean callbacks, any more than structured programming means gotos. Once a language actually has capabilities for handling new forms of programming (e.g. generators or async/await for async, or functions for structured programming) you're in a much better position to take advantage of the inherent power.

**Comment by rmg** [comment]
Node's async nature comes in large part from libuv and making heavy use of the existing *Sync functions already causes weird behaviour because it blocks the event loop.

Seems to me that to make a sync alternative to Node, you'd probably have to replace libuv before v8.

**Comment by bmeck** [comment]
@mvalente while interesting, this is probably not the right repository for this discussion. Many similar ideas have spawned around "sync" JS but javascript as a language never had concurrent shared memory and all of the implementations of such have had problems taking off because purely sync IO / scheduling does not match how people historically program Javascript. These have been tried many times and fail since JS programmers are somewhat uncomfortable with threads and synchronous concurrent programming at large scale. See:
- JXCore
- Vert.x
- SilkJS
- Ringo.js
- Rhino (Dyn.js now really)
- WScript (now that takes me back, and this is the worst for threads... ever...)

**Comment by bjouhier** [comment]
> Async doesn't mean callbacks, any more than structured programming means gotos.

Yes! Async programming can be painful in node. This is not a problem with the async model itself, it is a problem with the language: JS lacks an async/await construct. Once it gets it (ES7? @domenic probably knows better) async programming will be like a breeze. I've been doing it with my own async/await substitute (streamline.js) for more than 3 years and it's really pleasant. With this approach Java or .NET guys get up to speed with node in a few hours.

> If I had my way I'd remove everything with Sync in its name from the core API.

Yes, and in a hypothetical world I would also turn require async. This would make it possible to load code modules from a database, a web service, etc.

**Comment by medikoo** [comment]
I'd love to see some alternative for V8, even if it would be slower (but reasonably)

I think one of the biggest downsides of node is that its limits are dictated mostly by interpreter and not by hardware on which is run.

Currently if your process gets over 1.5Gb in memory, you need to think fast on optimisations or how to split it into multiple processes. It's not just about that it may crash soon, but because every simple operation (e.g. string search) becomes magnitudes slower. V8 is great for light processes, but otherwise seems not well suited.

Still it's just my impression, I'm not sure how other server languages compare in that.

**Comment by bnoordhuis** [comment]
> V8 is actually a huge pain in the backside to track for Node-core because of their development style

I don't disagree with that statement but SpiderMonkey and JSC are even worse.  At least the V8 team makes an effort of doing real releases and not breaking API in patch releases.

SpiderMonkey does a big code drop every now and then (last one is from December 2013);  JSC doesn't even do that, it's exclusively developed in-tree with WebKit.

**Comment by Fishrock123** [comment]
> And singlethreadedness hasnt stopped Apache and other webservers from being alternatives.

yeah um, because it does io **async** so that everything doesn't get clogged up.

**Comment by mvalente** [comment]
Great. Then nothing would stop a TraceMonkey/JS serverside of doing the
same. I could do
 async io and then you could use sync for the rest. Better yet: you could
have async/sync options
 for all language constructs. And people would opt. Even nicer if NodeJS
had the same thing.

On Fri, Oct 17, 2014 at 2:26 PM, Jeremiah Senkpiel <notifications@github.com

**Comment by bmeck** [comment]
@mvalente why not use SilkJS then?

**Comment by mvalente** [comment]
Indeed. Thats what I've been using. But it uses V8. I think that having a
Mozilla serverside JS
 using ***Monkey would be a nice option to have.

**Comment by mikeal** [comment]
While I don't find it all the interesting to do "Node but slightly different" there _are_ some interesting JavaScript projects that are going in a direction that supplant Node in the future like https://github.com/runtimejs

**Comment by heapwolf** [comment]
Runtime represents a different class of project though. It's more about infrastructure specialization, where as node is meant to run on general purpose systems.

**Comment by trevnorris** [comment]
@mvalente 

> JS is not singlethreaded.

What's meant by "JS is single threaded" is that you cannot share JS objects between heaps. Sure you can run multiple JS runtime instances in the same application, but you cannot in _any_ current gen engine pass a constructed JS object from one heap to another.

@bjouhier 

> Yes! Async programming can be painful in node. This is not a problem with the async model itself, it is a problem with the language: JS lacks an async/await construct.

This is completely subjective. I've been programming JS for 12 years now and using callbacks is so second nature to me that it requires zero additional cognitive reasoning.

**Comment by Qard** [comment]
It "requires zero additional cognitive reasoning" because you've been doing it for 12 years. People that are new to JS more often than not find the callback pattern tremendously awkward.

I can't even begin to count the number of times I've seen newbies write stuff like this:

```
app.get('/user/:id', function (req, res) {
  var user
  User.find(this.params.id, function (err, _user) {
    user = _user
  })
  res.json(user)
})
```

JS is very different from most languages. In some ways good, in others bad. But overall it is very difficult for newbies to get used to. Thing like generators, async/await, and promises try to reduce that learning curve.

**Comment by mvalente** [comment]
"requires zero additional cognitive reasoning" because you've been doing it
for 12 years.
"People that are new to JS more often than not find the callback pattern
tremendously awkward"

  Precisely. And that's a barrier do larger adoption. And it is the main
justification from those not adopting

**Comment by bmeck** [comment]
We need to be careful here. One of the reasons eventmachine and twisted struggle is because a _lot_ of ruby/python code is not non-blocking friendly. We can have async/await/generators make code read top to bottom without truly blocking and getting into the nightmares that those libraries suffer from. Merely adding sync for everything will cause major problems for async code (re: fibonacci hilarity).

If we have a proposal that comes up that does not encourage blocking synchronous code I would definitely be interested, but blocking synchronous code does not mesh with Node's ecosystem, or that of most JS environments.

**Comment by bjouhier** [comment]
@trevnorris I know this is a contentious issue. Some think that the callback model is natural. Others that it isn't. I need a bit of prose to explain my position so I'll do it on my blog.

@mvalente Sync APIs are a horrible fix. If you introduce sync APIs and you want to preserve concurrency you'll need threads. And if you have mutable state (this is JS, not Erlang) you'll need synchronization primitives. Now you'll be in a very different world: complex, inefficient (I've been there before). And you'll have a whole new slug of issues coming from the complex interactions between blocking threaded code and async code (@bmeck's comment above). 

A much better fix is to enhance the language so that your can code in sync style, while keeping the simple event loop and async model underneath.

Don't break the node model, just fix the language!

**Comment by Qard** [comment]
That is exactly why I like generators. You get the sync-like linearity, but
have yield keywords telling you where it's releasing control back to the
event loop.

Generally, within the context of a single request, there's not much that
can be done in parallel. It's really just switching between other requests
that node benefits from. And on the rare occasion you actually _can_ do
things in parallel, you can yield an array of tasks in koa.

**Comment by Fishrock123** [comment]
The main issue with generators and such is that they are inherently much slower than callbacks due to the lack of optimizations. Which brings me back to wondering if regenerator makes more optimizable code.

**Comment by bmeck** [comment]
@Fishrock123 just to clarify, they are not inherently slower (that would imply that they could not be as fast) just not optimized by the JIT right now.

**Comment by Fishrock123** [comment]
@bmeck I belive @trevnorris (& others) indicated they are inherently slower than callbacks.

I think the main reason is related to how they have to be internally structured.

(That is my understanding anyways, that being said generators are great.)

**Comment by domenic** [comment]
They are likely faster than callbacks as you can avoid generating a new call stack. In fact preliminary benchmarks show them to be faster than callbacks even without JIT.

**Comment by Fishrock123** [comment]
OH.

Ok whatever Dominic said. XD

**Comment by timoxley** [comment]
Most of the stuff people are requesting here will be enabled by ES6/ES7, which has very little to do with node. I advise to wait quietly for these features to land or use traceur to get them today.

Node's core team has indicated they will adopt new language features as they make sense. That discussion has been had a thousand times and there's little point discussing it again as the features required aren't even available yet. Note, without any drastic, opinionated changes to core, you can wrap it with whatever abstractions you want e.g. [visionmedia/co-fs](https://github.com/visionmedia/co-fs) i.e. if you want a different interface, just use a package or build one yourself that does what you want. This is likely going to be a better approach anyway – the APIs will be able to evolve faster and you get more fine-grained control over compatibility. Node should try to stay as vanilla as possible so such things can evolve in userland.

Making node easier for beginners by making node more like other platforms is not a particularly compelling argument. Neither JS or node have a problem with adoption, though they do have problems with people refusing to let go of legacy baggage they're lugging around from other ecosystems. Lack of async sugar doesn't actually prevent people from getting work done as there are a myriad of good solutions to augment async handling, available today, without requiring a single change to node.

**Comment by Mickael-van-der-Beek** [comment]
Seems strange to modify a language purely for adoption purposes.

It the end, even if sync methods would stay as training-wheels for newcomers, they would still be used by a good amount of the community in production just because they wouldn't know better.

The results would be counter productive because sure you'd have a larger user base but the overall performance of the apps produced would be inferior to having more difficult to grasp default concepts like async or generators.
This could create image issues where Node.js would be portrayed slow just because a percentage of the community uses the easy route in production.

Completely removing them would be the best choice imo.

Just my two cents as a nobody who uses Node.js.

**Comment by bjouhier** [comment]
@Mickael-van-der-Beek 
The performance of the apps produced would not be _inferior_, it would be _catastrophic_. This would not create an image of node being slow, it would create an image of node being broken.

You'd need threads to get viable performance levels out of Sync calls (talking about network apps of course, Sync calls are ok for command line utilities, or for app startup code).

**Comment by mvalente** [comment]
Those pesky webservers apache and nginx, getting away with catastrophic
performance all these years....

**Comment by bjouhier** [comment]
@mvalente They have threads!

**Comment by bjouhier** [comment]
@mvalente Answered too quickly: Apache has threads, nginx has an event loop and no Sync blocking calls.

**Comment by mvalente** [comment]
Indeed. That being so, a serverside JS engine can also be built using
threads and/or an event loop without the need for (mandatory) async
constructs

**Comment by Fishrock123** [comment]
> using threads

JS doesn't have "threads" that you can just "use" in any current VM that I know of

> without the need for (mandatory) async constructs

Considering that "sync" io blocks and that JS runs in a single threaded environment, yeah only if you want it to be awful.

It's better to teach people how to do things **correctly** for the language at hand.

Should we improve docs / guide / everything community and tutorials? **yes**
Should you build a fully "sync" javascript thing? **no**

EDIT:

Should there be node alternatives that use not-v8? probably, maybe?

**Comment by bmeck** [comment]
@mvalente Apache creates new run contexts (could be a thread, could be a CGI process, fast CGI process [which in turn gets to make its own run context], who cares) and then pipes data into that process. It is up to that context to created overlapped IO and has very little to do with apache's core. Also, not sure how this relates to mandatory async syntax when apache does not even mandate anything to the run context except that it will pipe in data as it gets it and read data as the context produces it (which may not be done synchronously prior to the end of the input data). NGINX does a similar approach except using an event loop rather than run context pooling (take note that various Apache tuning can turn it into async IO [use event mode instead of worker mode]).

Overall I am just curious why this matters since most people are not coding directly against the server internals.

**Comment by rlidwka** [comment]
> Should there be node alternatives that use not-v8? probably, maybe?

Yes. I'd definitely like SpiderMonkey alternative, because then I can try the latest ES6 features without any compilers. Or for comparison with v8, easily point out when certain node.js bug is caused by js engine for example.

I heard people work on this, see: https://bugzilla.mozilla.org/show_bug.cgi?id=1005411 , I'd like to see any news though.

**Comment by mvalente** [comment]
There are some misconceptions in some comments that would be too long too refute.

Fact: there are several alternatives for serverside JS that dont _demand_ async use: SilkJS, JSlibs, TeaJS, Persevere, GPSEE, etc, etc. There's been a module for Apache called mod_js for sometime. Hell there are some of us (me) that actually worked with SSJS in the early 90s using Netscape's Livewire. 

Why not use one of them, right? Because of potential risk of work on them being discontinued. A version by Mozilla would solve for that. And aditionally bring credibility.

Does it have to be sync only? No. Does it have to be async only? No. Would it be nice to have the option of choosing sync/async? Yes. Node forces the use of async without the need for it (see above). Would it be great to have a SSJS option coming from Mozilla that focused more on sync? Yes. Would it need to be sync only? No, it could have async primitives, but main focus on sync. Why? Because there are a lot of programmers out there coming from PHP, Python, Perl that find the functional/async programming paradigm to be too difficult and when trying to learn it go into the dreaded callback hell.

SSJS didnt show up in the last couple of years https://developer.mozilla.org/en-US/docs/Archive/Web/Server-Side_JavaScript/Walkthrough 

And Node didnt show up in the last couple of years as well http://mvalente.eu/2009/11/25/requirements-for-a-modern-web-development-framework/

Or in 2008
http://www.infoq.com/news/2008/03/tsa-working-group-site
http://goo.gl/PYZY35

Jus my 2 cents. I would like to have a sync paradigm SSJS solution. Preferably coming from Mozilla.

**Comment by mikeal** [comment]
> JS doesn't have "threads" that you can just "use" in any current VM that I know of

There actually has been experimentation using v8 isolates in a "thread like manor" for sync style network operations. None of them really took off because it's much slower than the approach Node has taken.

**Comment by mikeal** [comment]
> Node's core team has indicated they will adopt new language features as they make sense. 

This is slightly misleading. The core team has said on many occasions that they will maintain a lack of opinion about the maturity of language features and adopt them as they land in v8. If they land in v8 with flags, they are in node with flags. If they land without special flags, they'll be in without special flags. I don't think anyone has ever advocated that features available in v8 be kept from Node users.

This should be enough for the community to adopt them. Node's core APIs won't be adopting new patterns until long after they are proven in the community, if ever. That's a good thing, some of the most troublesome parts of Node are things we adopted too fast and now wish were just modules.

The biggest concern keeping people from these features isn't "core adoption," it's not even v8 adoption, it's a lack of consistent releases that take new versions of v8 as they implement new features.

**Comment by mvalente** [comment]
One aditional comment regarding the "mandatory" use of async/callback primitives from some recent zealot NodeJs converts: async/callbacks are a way of implementing messaging in programming. There are other messaging options like events, pubsub, promises streams. Personally I prefer events and pubsub.

**Comment by greelgorke** [comment]
why is this even a discussion here? you want to go sync-style? go for it, there are plenty of options. node.js core api is as low level as it makes sense, and probably should be even lower level. there is the most primitive async model implemented, which can be wrapped into higher lvl constructs. just do it.

if some ppl find async difficult, then point them to the right tool for them. so whats wrong with that? sync+threaded style isn't easy either, why is learning effort even a fact?

**Comment by bmeck** [comment]
@mvalente it sounds like you want:
1. SpiderMonkey instead of V8
2. Sync instead of Async
3. Threads instead of EventLoop
4. An IO multiplexer instead of managing IO (Apache comment?)

I am not sure this even sounds like Node, it is comparable to telling PHP/python/perl to move all sync stuff to allow async and change out the underlying IO and runtime.

While some argument could be made for PHP/python/perl programmers having difficulty with Node's paradigms we should remember that Node was created in an opinionated manner and has not had problems with adoptions (quite the opposite). I feel it would be outrageous (literally) to change all of the underlying and move away from JavaScript convention and design patterns to satisfy an adoption concern that simply is not present. The number of hobbyist and enterprise developers using node not just as an HTTP server but as a general tool makes adoption a hard sell. The sheer amount of code academy and acceptance of JS for almost all client side code also makes the difficulty to learn a hard sell.

**Comment by mvalente** [comment]
"you want to go sync-style? go for it, there are plenty of options"

Yes. All without "relevant/credible" backing that would assure it would be available long term.

"SpiderMonkey instead of V8"
Yes. Preferably.

"Sync instead of Async"
Yes.

"Threads instead of EventLoop"
Not necessarily. But yes.

"An IO multiplexer instead of managing IO"
Yes.

"I am not sure this even sounds like Node"
Precisely. Thatś why I would like an alternative.

" I feel it would be outrageous (literally) to change all of the underlying"
I'm not arguing for that. I'm arguing that, since Node is very unlikely to provide sync primitives (as we can see from this discussion, it has become a religion war like emacs/vi), I just think that an alternative to Node would be to provide a sync implementation of the primitives. 

"sheer amount of code academy and acceptance of JS for almost all client side code"
I would argue that up until a couple of years ago (and even today) client side JS mostly used events for messaging and the use of callbacks were/are rare. I know long time JS programmers that find it difficult to get into the async/callbacks mindframe.

**Comment by sam-github** [comment]
Its reasonable to want an alternative to node, but why is this conversation here? How is building a not-node relevant to moving node forward?

**Comment by Fishrock123** [comment]
I don't think it is. This should probably be closed. (& locked maybe)

**Comment by greelgorke** [comment]
@mvalente then just go for vert.x or something. really, no offense, but what you're asking for sonds not like a node other way, but an other tool

**Comment by jamlfy** [comment]
@Fishrock123 Saying that V8 is faster, it is not true. Far from the versions the latest versions of mozilla. That is a myth!

I share the same position @bnoordhuis, which is chosen at the beginning V8 to have very specific characteristics, fine. However I think it is time to assess whether it is feasible to proceed maintained V8 engine as NodeJS.

**Comment by Fishrock123** [comment]
@alejonext it **was** faster at the time.

**Comment by rlidwka** [comment]
Yep, v8 was a couple of times faster than everything else at the time. And it played a big part in node.js success. There were other server-side js wrappers, but they were just too slow to matter.

But recent news from mozilla makes me wish we'd go for dual-engine support when people can choose which engine to use.

**Comment by jamlfy** [comment]
@rlidwka  would be a similar model to this .. 

```
 +-----------------------+
 |   NPM (node_modules)  |
 +-----------------------+
 |      NodeJs Core      |
 +-----------------------+
 |     V8    |  Mozilla  |
 +-----------------------+
```

**Comment by othiym23** [comment]
While it would be possible to write a platform abstraction layer for Node to abstract away the differences between V8 and SpiderMonkey's embedding APIs, my totally seat-of-the-pants guess is that that code would be roughly comparable in complexity to libuv, which is to say a significant and complex project all its own. Supporting one JS engine (regardless of whether it's V8 or SpiderMonkey) is much easier than supporting two.

**Comment by Fishrock123** [comment]
> ... or implement V8 api on top of SpiderMonkey

this has to be the most crazy idea I've heard all week haha, what with v8's pace and all.

idk maybe it'l work

**Comment by vkurchatkin** [comment]
@rlidwka node is really coupled with v8, so you have to implement all of v8 API, which is pretty much impossible

**Comment by jamlfy** [comment]
@vkurchatkin I do not think it impossible, Nor do I believe that it is so complex. In fact the current NodeJS code only uses things done in C for file system and others. For the rest, all l used nodejs layer and its libraries.

**Comment by piscisaureus** [comment]
I am very much in favor of supporting multiple js engines. V8 is great indeed but we don't know (and right now we can't know) if it's actually the best for typical node.js workloads.

Also, avatar.js is already a node implementation on top of a different js engine (nashorn). I don't see why spidermonkey (and javascriptcore, chakra) could not be supported. But I do realize that fork lifting node onto another engine is a lot of work, and without leniency from the node project it might never happen.

The question is how to go about it. Creating a v8-api wrapper around spidermonkey and other engines as some have suggested may seem an obvious solution, but  you'll run into corner cases pretty fast. That's because you now need to map the concepts and abstractions that the "other" engine provides to v8 concepts exactly. About 3 years ago mozilla tried it (see [v8money](https://github.com/zpao/v8monkey)), but that went nowhere. Also, the v8 api is a moving target. 

It seems that blessing a project like nan (and using it to implement core bindings) has more potential; nan can choose it's own API so it would be possible to define concepts that can be easily mapped to different engines.

I've actually done some investigative work in this area not so long ago, in an attempt to find "common ground" between v8, spidermonkey and chakra. I found that these were the most difficult issues to deal with:

### GC-rooting javascript values

The v8 API allows one to root values globally with `Persistent<T>` and within a function scope using `Local<T>`. 
Spidermonkey nowadays has an API that's very similar (see their [GC Rooting Guide](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/GC_Rooting_Guide)) .
Chakra offers reference counting only (see [JsAddRef](http://msdn.microsoft.com/en-us/library/dn249659%28v=vs.94%29.aspx) and [JsRelease](http://msdn.microsoft.com/en-us/library/dn249647%28v=vs.94%29.aspx)).

### Attaching externally managed memory to a javascript object

Sometimes node needs to associate externally managed memory with a javascript object.
Prime examples are `Handle`s, `Req`s and `Buffer`s in node-core. `Handle`s and `Req`s have a c++ object instance attached to them which contains information about an open libuv handle and an pending asynchronous operation, respectively. `Buffer`s are attached to their backing store that lives outside of javascript heap.

v8 has the `SetAlignedPointerInInternalField` API that lets you associate any number of pointers with a v8 object.
Same goes for spidermonkey, which has `js_SetReservedSlot` and `js_getReservedSlot`.
Chackra supports only one pointer per object through `JsSetExternalData`. This is sufficient for node.

### Garbage collector notification

Continuing the previous point - external data requires integration with the garbage collector. For example node frees the backing store for `Buffer` when the garbage collector notifies that the associated javascript value has been GCed. There are other types of objects that use a weak callback (e.g. FSWatcher); I believe this is poor design, but that's out of scope for this exercise.

v8 supports this with by providing a "weak callback" for objects that have been made weak. The v8 API is "overkill" for node because it allows the weak callback to resurrect the object, but node never does this.
I haven't figured out yet how to do it with spidermonkey.
Chakra has `JsSetRuntimeBeforeCollectCallback()` which is similar to v8's weak callback except that it doesn't support resurrection. 

### The vm needs direct access to external buffer memory

As an extension to the previous point - to support an efficient `Buffer` implementation, the javascript engine must be able to directly use a binary data that lives outside of the javascript heap. 

V8 supports this through the `SetIndexedPropertiesToExternalArrayData` api. 
Spidermonkey does support something similar through `CreateTypedArrayWithBuffer`, but that (obviously) creates a typed array and not a Buffer object.
Chakra doesn't support this as far as I can tell.
I recall from speaking with one of the avatar.js developers about half a year ago that this was a problem for them too, because at the time nashorn had no such api. 

Note that technically it is not necessary for the `Buffer` backing store to live outside of the javascript heap; however it must not move, libuv may be using that buffer at the same time to perform an asynchronous read or write. The v8 memory manager moves objects around, and I expect other engines to do the same (or at least reserve the right to do so).

**Comment by vkurchatkin** [comment]
In JSC:
- GC-rooting : [JSValueProtect](https://developer.apple.com/library/mac/documentation/JavaScriptCore/Reference/JSValueRef_header_reference/index.html#//apple_ref/c/func/JSValueProtect), [JSValueUnprotect](https://developer.apple.com/library/mac/documentation/JavaScriptCore/Reference/JSValueRef_header_reference/index.html#//apple_ref/c/func/JSValueUnprotect)
- External memory : [JSObjectSetPrivate](https://developer.apple.com/library/mac/documentation/JavaScriptCore/Reference/JSObjectRef_header_reference/index.html#//apple_ref/c/func/JSObjectSetPrivate)
- GC notifications : [JSClassDefinition.finalize](https://developer.apple.com/library/mac/documentation/JavaScriptCore/Reference/JSObjectRef_header_reference/index.html#//apple_ref/c/tdef/JSClassDefinition), for host objects only
- External array data : nothing like that

**Comment by piscisaureus** [comment]
@vkurchatkin Thanks. On external array data, is there no API to create an uint8 TypedArray?

**Comment by vkurchatkin** [comment]
@piscisaureus no, API is es5 only. You can evaluate `new Uint8Array(%d)`, but no access to underlying buffer (as in v8).

Technically implementing buffer-like object is not a problem, we need a `JSClass` with getter callback (and a bunch of others), which dereferences private pointer and looks up data cell. Data can be freed in `finalize` callback.

The problem is: `JSClass` is used to create object with C API, and it appears to be no good way to turn `JSClass` into constructor.

**Comment by piscisaureus** [comment]
@vkurchatkin Thanks! That sounds like an acceptable workaround, although I think that turning every indexed lookup into a virtual function call won't be very performant. I wonder how Safari implements webgl though if the js api doesn't allow buffer access.

**Comment by vkurchatkin** [comment]
@piscisaureus this API is for embedders, webkit doesn't use any of it, as far as I know

**Comment by domenic** [comment]
I would think that, if people were given the feature of supporting multiple JS engines, they might be willing to pay the cost of moving from non-standard `Buffer`s to standard typed arrays? The reason we still have Buffers is just back-compat, right? I don't remember them being technically superior, apart from being a great solution before JS typed arrays were properly standardized and implemented.

**Comment by vkurchatkin** [comment]
@domenic one more advantage: Buffers use uninitialized memory. It would be cool if spec made initialization optional (or a part of Annex B)

**Comment by domenic** [comment]
Oh, right, I forgot about that :(. Pretty hard to imagine that kind of security hole being put in the spec. Maybe some optimizations could be in place though so that it gets zeroed out just in time for use or something? Bleh, I dunno.

**Comment by vkurchatkin** [comment]
@domenic why is this a security hole? I mean, for browsers, it is, but ECMAScript is more than just a language for web scripts. This requirement can be put in Annex B or https://javascript.spec.whatwg.org/

**Comment by domenic** [comment]
Anything that allows you to read arbitrary memory is a security hole in any environment.

**Comment by vkurchatkin** [comment]
@domenic I don't get it. So `Buffer` is a security hole? or `malloc`?

**Comment by sergioramos** [comment]
> About 3 years ago mozilla tried it (see v8money), but that went nowhere. 

https://twitter.com/zpao/status/468860127391809536

**Comment by mikeal** [comment]
v8 has made breaking changes in their API enough times now that we MUST assume it will break again, and again, and probably again. simulating an API that is by any definition unstable seems like a bad idea.

**Comment by bnoordhuis** [comment]
> I don't get it. So Buffer is a security hole?

Not exactly "is" but "can be".  Imagine you store confidential data like a password in memory.  When the memory is (re)used by a buffer and you have a bug in your application that transmits the buffer over the network without clearing it first, you leak confidential data.

> v8 supports this with by providing a "weak callback" for objects that have been made weak. I haven't figured out yet how to do it with spidermonkey.

With SM, it's the JSClass.finalize callback of the object's backing JSClass.

> v8 has the SetAlignedPointerInInternalField API that lets you associate any number of pointers with a v8 object.  Same goes for spidermonkey, which has js_SetReservedSlot and js_getReservedSlot.

FWIW, I could never get this to quite match the V8 semantics, even with the undocumented JSClass magic for making an object indexable (which was super slow, by the way.)

EDIT: IIRC, the issue was that `SlowBuffer.prototype.__proto__ === Buffer.prototype` in v0.10, something I couldn't express with SM's class system.  Perhaps not an issue with v0.12.

**Comment by vkurchatkin** [comment]
@bnoordhuis if you have a bug, everything can be a security hole: filesystem operations, child processes, etc. Anyway, I understand HOW this can be an issue, I just don't think it should be a part of ECMAScript spec, because it is not limited to sandboxed environments.

**Comment by apaprocki** [comment]
Hi, I figured I'd chime in about what we do at Bloomberg...

Re: multi-engine binding -- that is essentially what we have been doing for many years. We have an engine independent set of binding interfaces and as of right now, we have concrete implementations for V8, SM, and even Lua (to just keep us honest that it really does work abstractly for dynamic Object-like things). Our internal JS process is not a Node replacement -- we had very different goals from what Node itself provided. This binding layer is just one of the many building blocks one could use to build a Node equivalent. We write nice native modules once and they can be used in any engine that has a concrete adapter. We currently run our JS process atop either V8 or SM depending on what hardware architecture it is running on.

I want to try to get at least the binding layer open-sourced up on our GH account, but it depends on a few other types/components in our stack that are not yet available in http://github.com/bloomberg/bde. I'll continue working on getting the prerequisites released, but I can't be sure when it will happen.

**Comment by mvalente** [comment]
Just my 2cents https://github.com/olegp/common-node

**Comment by trevnorris** [comment]
@piscisaureus 

> Note that technically it is not necessary for the Buffer backing store to live outside of the javascript heap; 

Disagree. It'd be really dumb to clog up the heap with large objects. It might be a corner case, but still very useful when I can load large data sets into memory and work on them from memory.

@domenic 

> The reason we still have Buffers is just back-compat, right?

Impossible to switch to typed arrays because we can't extend natives, meaning we can't allow both quick indexed array access and additional methods on the same object.

@bnoordhuis 

>  IIRC, the issue was that `SlowBuffer.prototype.__proto__ === Buffer.prototype` in v0.10, something I couldn't express with SM's class system. Perhaps not an issue with v0.12.

That's no longer an issue in v0.12.

**Comment by piscisaureus** [comment]
@trevnorris 

> Disagree. It'd be really dumb to clog up the heap with large objects. It might be a corner case, but still very useful when I can load large data sets into memory and work on them from memory.

That's definitely true for v8 because it has a fully managed heap with a size limit.
However other engines may just malloc() space for javascript objects, so it doesn't really matter whether the vm or node manages that allocation - the net result is the same.
But I agree we shouldn't haphazardly assume that buffers can live on a managed heap.

**Comment by domenic** [comment]
> Impossible to switch to typed arrays because we can't extend natives, meaning we can't allow both quick indexed array access and additional methods on the same object.

When I say "switch," I mean in a non-API-compatible way. Given that, what additional methods do you need (that you can't get from e.g. a DataView)?

**Comment by piscisaureus** [comment]
@domenic 

```
> Buffer
{ [Function: Buffer]
  isEncoding: [Function],
  poolSize: 8192,
  isBuffer: [Function: isBuffer],
  byteLength: [Function],
  concat: [Function] }
> Buffer.prototype
{ inspect: [Function: inspect],
  get: [Function: get],
  set: [Function: set],
  write: [Function],
  toJSON: [Function],
  toString: [Function],
  fill: [Function: fill],
  copy: [Function],
  slice: [Function],
  utf8Slice: [Function],
  binarySlice: [Function],
  asciiSlice: [Function],
  utf8Write: [Function],
  binaryWrite: [Function],
  asciiWrite: [Function],
  readUInt8: [Function],
  readUInt16LE: [Function],
  readUInt16BE: [Function],
  readUInt32LE: [Function],
  readUInt32BE: [Function],
  readInt8: [Function],
  readInt16LE: [Function],
  readInt16BE: [Function],
  readInt32LE: [Function],
  readInt32BE: [Function],
  readFloatLE: [Function],
  readFloatBE: [Function],
  readDoubleLE: [Function],
  readDoubleBE: [Function],
  writeUInt8: [Function],
  writeUInt16LE: [Function],
  writeUInt16BE: [Function],
  writeUInt32LE: [Function],
  writeUInt32BE: [Function],
  writeInt8: [Function],
  writeInt16LE: [Function],
  writeInt16BE: [Function],
  writeInt32LE: [Function],
  writeInt32BE: [Function],
  writeFloatLE: [Function],
  writeFloatBE: [Function],
  writeDoubleLE: [Function],
  writeDoubleBE: [Function] }
```

**Comment by domenic** [comment]
@piscisaureus yes, ArrayBuffer combined with DataView gives you all of those... except maybe not the little/big endian distinction, and you have to use text encoding APIs for the text stuff.

**Comment by trevnorris** [comment]
@domenic 
`fill()`? Also, coming in v0.12 is a generic `(Read|Write)(U)Int(LE|BE)()` that can handle 8, 16, 24, 32, 40 and 48 bit reads/writes.

`TextDecoder` has several pit falls. First is that it doesn't support `ISO-8859-1` (AFAIK). Another is the API. Let's take a quick look at creating a buffer of arbitrary length, filling it with character values and finally reading an arbitrary section of it (doing this to the best of my knowledge of existing specs):

First Node:

``` javascript
// size, start, end ∈ ℕ; start + end <= size
// fill is a number or string
function genString(size, fill, start, end) {
  var b = new Buffer(size).fill(fill);
  return b.toString(start, end);
}
```

Now using browser APIs:

``` javascript
// size, start, end ∈ ℕ; start + end <= size
// fill is a number or string
function genString(size, fill, start, end) {
  var dv = new DataView(new ArrayBuffer(size));
  if (typeof fill === 'number')
    fillDVNumber(dv, fill);
  else
    fillDVString(dv, fill);
  var td = new TextDecoder('utf-8');
  return td.decode(new DataView(td.buffer, start, end));
}

function fillDVNumber(dv, num) {
  for (var i = 0; i < dv.byteLength; i++)
    dv.setUint8(i, num % 0xff);
}

function fillDVString(dv, str) {
  // not going to take the time implementing a utf-8 safe
  // string fill.
}
```

There are other small pains of things like needing to externalize an `ArrayBuffer` before external APIs can directly access its memory, and that I can't allocate uninitialized memory.

Additionally, the `Buffer` implementation of numeric `read`/`write` methods are much faster than those for `DataView()` (~8ns vs ~47ns).

Granted there are some parts of the Typed Array API I really like, but it just isn't well suited to working with data the way Node does.

**Comment by domenic** [comment]
@trevnorris re: fill, http://people.mozilla.org/~jorendorff/es6-draft.html#sec-%typedarray%.prototype.fill

But, in general, thanks for the detailed comparison. I still would hope that if we were starting from a world in which typed arrays already existed, we wouldn't invent Buffers. The uninitialized-memory thing seems like the only potential actual dealbreaker.

**Comment by trevnorris** [comment]
@domenic Thanks for the link to the specification.

One difference between that `fill()` implementation and what's coming in v0.12 is that:

``` javascript
var b = Buffer(8).fill('abcd');
console.log(b.toString());
// output: 'abcdabcd'
```

Minor, but still useful.

Also, was I mistaken that `TextDecoder` doesn't support `ISO-8859-1`? I hope it does.

**Comment by domenic** [comment]
> Minor, but still useful.

Yeah, in general it seems like Buffer has more ergonomic text-manipulation stuff.

> Also, was I mistaken that TextDecoder doesn't support ISO-8859-1? I hope it does.

You can decode iso-8859-1 (aka windows-1252, and [lots more](https://encoding.spec.whatwg.org/#names-and-labels)), but you can only encode as UTF-8 or UTF-16.

**Comment by trevnorris** [comment]
> You can decode iso-8859-1 (aka windows-1252)

If that were the case then I wouldn't expect the following:

``` javascript
var b = new Buffer(1);
b[0] = 159;
// 'binary' is V8's one-byte encoding (ISO-8859-1)
b.toString('binary');
// output: ''
// expected: 'Ÿ'
```

That might just be a V8-ism, and is possible it changes. Just FYI.

> but you can only encode as UTF-8 or UTF-16.

Ah, that's painful. Know if there's any discussion about extending that in the future?

**Comment by domenic** [comment]
> That might just be a V8-ism, and is possible it changes.

Yeah, not sure what relation V8's encoding has to windows-1252/iso-8859-1/etc.

> Know if there's any discussion about extending that in the future?

Not for the web; the proliferation of non-UTF-8 text is harmful (I'm hoping we can remove UTF-16 actually...) and shouldn't be convenient. You can make the argument that it's more necessary for a server-side runtime I guess, but then again you can make the same argument that non-UTF-8 encodings are harmful and should be inconvenient (i.e. require user-created encoders).

**Comment by piscisaureus** [comment]
> TextDecoder has several pit falls. First is that it doesn't support ISO-8859-1 

Well node doesn't either. The "binary" encoding is effectively latin-1.

**Comment by piscisaureus** [comment]
@domenic @trevnorris I think it's be better to move TypedArray/Buffer talk to a separate issue

**Comment by jamlfy** [comment]
In conclusion we have problems with the buffer and the character encoding? To construct an alternative

**Comment by 5nyper** [comment]
An alternative to nose could be Dart, benchmarks prove the DartVM is faster than V8

**Comment by rlidwka** [comment]
> An alternative to nose could be Dart, benchmarks prove the DartVM is faster than V8

I surely hope it's a joke. If not, I'll happily suggest to switch to Erlang instead. :)

**Comment by feross** [comment]
Yo guys, let me tell you about Haskell... It's just like node (except not
really...)

**Comment by 5nyper** [comment]
No it isn't a joke, what I HOPE that your reply is a joke.

**Comment by Fishrock123** [comment]
looks like we are switching to erlang guys

**Comment by 5nyper** [comment]
Take a look at this then:  http://www.infoq.com/news/2013/05/Dart-Java-DeltaBlue

**Comment by Fishrock123** [comment]
Ok, node is named node.[javascript] for a reason. Quite frankly there are many other languages already out there for I/O (go, rust, java, etc).

I think anything that isn't about node as a javascript I/O is definitely off-topic. (Not that this thread hasn't been for a long time.)

**Comment by kav** [comment]
It seems that the primary confusion here is this: Node was designed as a platform for writing async server applications that happens to use javascript.

Many folks seem to think node is for writing server applications in javascript and happens to be async. 

Thus they confuse the urge to remove problematic Sync constructs as zealotry when it's a totally rational move for a non-blocking server environment. Conversely they can't figure out why everyone gives them looks when they want to add a solid synchronous model. Which does make perfect sense if you just want one language on your client and server.

Node is non-blocking and async first and foremost and uses javascript secondarily.  Adding a bunch of sync stuff thus doesn't really make sense from a fundamental standpoint.

This does seem to come up an awful lot. More and more as big companies join the party. Does anyone have a great resource to point to that explains the what and whys of node? Might help explain the crazy looks and perception of zealotry.

TL;DR
It'd be easier to drive planes if they didn't have wings and go up in the air. But then we'd call them cars, not planes.

**Comment by domenic** [comment]
> Node is non-blocking and async first and foremost and uses javascript secondarily.

Asserting this does not make it true.

**Comment by mvalente** [comment]
And then there are people who have been on the JS serverside camp for
more
 than 6 years and who find it funny to see displays of crazy looks and
zealotry which
 are actually proper for the newly converted.

   http://mvalente.eu/?s=Tracemonkey&x=13&y=12

**Comment by Fishrock123** [comment]
I hear io.js is a good alternative to node.js

**Comment by erikcorry** [comment]
@medikoo If you can boil down your large-heap performance issues to a reasonable repro, then the V8 team are very interested in bug reports.  The GC of V8 is not as scalable as the JVM's, which is said to scale to dozens of threads and 100s of Gbytes, but 1.5Gbyte should be no problem.

Note that you may need to up the sizes of various limits.  Primarily --max-old-space-size and --max-executable-size .  You may want to experiment with --nouse-idle-notification too.

**Comment by medikoo** [comment]
@erikcorry great thanks for that valuable information. I'll try to investigate that further and eventually I'll come up with some solid test case

**Comment by bnoordhuis** [comment]
@medikoo Just a heads up, --nouse-idle-notification is a no-op in v0.10 and newer; I removed idle notifications in iojs/io.js@d607d856.

**Comment by kaizhu256** [comment]
synchrounous io is legitimate in one-time calls like initializing server configuration and models.  after spending 2 years as an async zealot, I've come to realize its very inefficient and brittle for writing nontrivial init code.

**Comment by mvalente** [comment]
Oh, looky here! Building a real alternative to Node using Spidermonkey is actually possible, contrary to the naysayers...

https://github.com/jxcore/jxcore

http://jxcore.com/home/

**Comment by erikcorry** [comment]
Nice try, but I know what the date is today!

**Comment by mvalente** [comment]
:-)

**Comment by Fishrock123** [comment]
> https://github.com/jxcore/jxcore

Cool. That's good, there should be options.

Our business is node/io.js. If we have the resources to one day make the js engine swappable, that'd be great. For now though, that simply isn't the case.

@mikeal Can you lock this?
